sklearn/ensemble/tests/test_forest.py

```diff
@@ -1330,3 +1330,98 @@ def test_forest_degenerate_feature_importances():
     gbr = RandomForestRegressor(n_estimators=10).fit(X, y)
     assert_array_equal(gbr.feature_importances_,
                        np.zeros(10, dtype=np.float64))
+
+
+@pytest.mark.parametrize('name', FOREST_CLASSIFIERS_REGRESSORS)
+@pytest.mark.parametrize(
+    'max_samples, exc_type, exc_msg',
+    [(int(1e9), ValueError,
+      "`max_samples` must be in range 1 to 6 but got value 1000000000"),
+     (1.0, ValueError,
+      r"`max_samples` must be in range \(0, 1\) but got value 1.0"),
+     (2.0, ValueError,
+      r"`max_samples` must be in range \(0, 1\) but got value 2.0"),
+     (0.0, ValueError,
+      r"`max_samples` must be in range \(0, 1\) but got value 0.0"),
+     (np.nan, ValueError,
+      r"`max_samples` must be in range \(0, 1\) but got value nan"),
+     (np.inf, ValueError,
+      r"`max_samples` must be in range \(0, 1\) but got value inf"),
+     ('str max_samples?!', TypeError,
+      r"`max_samples` should be int or float, but got "
+      r"type '\<class 'str'\>'"),
+     (np.ones(2), TypeError,
+      r"`max_samples` should be int or float, but got type "
+      r"'\<class 'numpy.ndarray'\>'")]
+)
+def test_max_samples_exceptions(name, max_samples, exc_type, exc_msg):
+    # Check invalid `max_samples` values
+    est = FOREST_CLASSIFIERS_REGRESSORS[name](max_samples=max_samples)
+    with pytest.raises(exc_type, match=exc_msg):
+        est.fit(X, y)
+
+
+@pytest.mark.parametrize('name', FOREST_CLASSIFIERS)
+def test_classification_toy_max_samples(name):
+    # Test that the toy example is separable via a bootstrap size of only 2
+
+    rng = np.random.RandomState(1)
+    max_tries = 100
+
+    # The toy example is separable using just one
+    # decision stump, and choosing 2 examples from the full
+    # 6-example dataset *if* the 2 examples are chosen correctly.
+    est = FOREST_CLASSIFIERS[name](
+        n_estimators=1,
+        bootstrap=True,
+        max_samples=2,
```

Comment:
Would this test fail if max_samples=None? If not, I don't get what it's testing

Reply:
Yea, it's a bit awkward. I wrote it prior to the suggestion of checking the node count. The intent was that this would be a sanity check for correctness when `max_samples` is provided, rather than expecting failure when `max_samples=None`. Anyway, I'll remove it.